ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java

```diff
@@ -116,9 +116,11 @@ protected int getPartitions(final KafkaTopicClient kafkaTopicClient) {
   public Stream<ColumnName> resolveSelectStar(
       final Optional<SourceName> sourceName, final boolean valueOnly
   ) {
-    return getSources().stream()
-        .filter(s -> !sourceName.isPresent() || sourceName.equals(s.getSourceName()))
-        .flatMap(s -> s.resolveSelectStar(sourceName, false));
+    return getSources()
+            .stream()
+            .flatMap(s -> s instanceof JoinNode ? s.getSources().stream() : Stream.of(s))
```

Comment:
This feels wrong - can we now rely on polymorphism here? i,.e. can we not simply call the source nodes and have then do the right thing, rather than having these `instanceof` conditionals?

Reply:
What I want to achieve is very specific to nested Join nodes. That means I have three ways I see to solve this:

- use `instanceof` and just flatten the sources here (existing solution)
- add a new method `getLeafSources` (seems like overkill)
- change `getSources()` to recurse (I tried this and it didn't work out of the box, so it requires more changes)

I think the third option makes the most sense, I'll do that in a follow-up PR since it wasn't a one-shot change.

Reply:
EDIT: `getSources()` shouldn't recurse because it should give you exactly _this_ node's sources. so it's between adding a new method and keeping it as is. I think I like it as is, but let me know if you feel strongly

Reply:
Don't feel strongly about it - just feels wrong. Without bringing the code down and trying out your changes I don't know a better way - so happy to merge.